app/text_buffer.py

```diff
@@ -138,155 +141,223 @@ def doMoveLines(self, begin, end, to):
+          self.doMoveLines(begin, end, to)
+        elif change[0] == 'n':
+          # Redo split lines.
+          line = self.lines[self.penRow]
+          self.lines.insert(self.penRow + 1, line[self.penCol:])
+          self.lines[self.penRow] = line[:self.penCol]
+          for i in range(max(change[1][0] - 1, 0)):
+            self.lines.insert(self.penRow + 1, "")
+        elif change[0] == 'v':  # Redo paste.
+          self.insertLines(change[1])
+        elif change[0] == 'vb':
+          self.penCol -= len(change[1])
+          row = min(self.markerRow, self.penRow)
+          rowEnd = max(self.markerRow, self.penRow)
+          for i in range(row, rowEnd + 1):
+            line = self.lines[i]
+            x = self.penCol
+            self.lines[self.penRow] = line[:x] + line[x + len(change[1]):]
+        elif change[0] == 'vd':  # Redo vertical delete.
+          upperRow = min(self.markerRow, self.penRow)
+          lowerRow = max(self.markerRow, self.penRow)
+          x = self.penCol
+          for i in range(upperRow, lowerRow + 1):
+            line = self.lines[i]
+            self.lines[i] = line[:x] + line[x + len(change[1]):]
+        elif change[0] == 'vi':  # Redo vertical insert.
+          text = change[1]
+          col = self.penCol
+          row = min(self.markerRow, self.penRow)
+          rowEnd = max(self.markerRow, self.penRow)
+          app.log.info('do vi')
+          for i in range(row, rowEnd + 1):
+            line = self.lines[i]
+            self.lines[i] = line[:col] + text + line[col:]
+        else:
+          app.log.info('ERROR: unknown redo.')
+      # Redo again if there is a move next.
+      if (self.redoIndex < len(self.redoChain) and
+          self.redoChain[self.redoIndex][0] == 'm'):
+        self.redo()
 
   def redoAddChange(self, change):
     """Push a change onto the end of the redoChain. Call redo() to enact the
         change."""
+    def noOpInstruction(change):
+      # Eliminate no-op entries
+      # app.log.info('opti', change)
+      noOpInstructions = set([
+        ('m', (0,0,0,0,0)),
+      ])
+      assert ('m', (0,0,0,0,0)) in noOpInstructions
+      if change in noOpInstructions:
+        return True
+      return False
+
     if self.debugRedo:
       app.log.info('redoAddChange', change)
     # When the redoChain is trimmed we may lose the saved at.
-    if self.redoIndex < self.savedAtRedoIndex:
-      self.savedAtRedoIndex = -1
-    self.redoChain = self.redoChain[:self.redoIndex]
+    # Trim only when there is a non-trivial action.
+    if change[0] == 'm' and self.redoIndex != len(self.redoChain):
+      self.dirtyChain = True
+    else:
+      # Trim and combine main redoChain with dirtyChain
+      # if there is a non-trivial action
+      if self.redoIndex < self.savedAtRedoIndex:
+        self.savedAtRedoIndex = -1
+      self.redoChain = self.redoChain[:self.redoIndex]
+      if len(self.tempChain) and self.redoChain[-1][0] == 'm':
```

Comment:
Try: 
1. make an edit
2. undo the edit (so the redoChain is empty)
3. move cursor (so tempChain is not empty)
4. press backspace (or something to get here)
problem: exception on an index error on the redoChain.